timers/timers.go

```diff
@@ -0,0 +1,164 @@
+package timers
+
+import (
+	"runtime"
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/uber-go/atomic"
+)
+
+const (
+	cacheline = 64
+
+	// State machine for timers.
+	stateDefault = iota
+	stateExpired
+	stateCanceled
+)
+
+type mutex struct {
+	sync.Mutex
+	// Avoid false sharing:
+	// http://mechanical-sympathy.blogspot.com/2011/07/false-sharing.html
+	pad [cacheline - unsafe.Sizeof(sync.Mutex{})]byte
+}
+
+// Timer FIXME
+type Timer struct {
+	f        func()
+	deadline uint64 // in ticks
+	state    atomic.Int32
+	prev     *Timer
+}
+
+// Stop FIXME
+func (t *Timer) Stop() {
+	t.state.Store(stateCanceled)
+}
+
+func (t *Timer) append(tail *Timer) *Timer {
+	tail.prev = t
+	return tail
+}
+
+// Wheel FIXME
+type Wheel struct {
+	period time.Duration
+	ticker *time.Ticker
+
+	locks      []*mutex
+	lockMask   uint64
+	buckets    []*Timer
+	bucketMask uint64
+
+	fired   chan *Timer
+	stop    chan struct{}
+	stopped chan struct{}
+}
+
+// NewWheel creates and starts a new Wheel.
+func NewWheel(period time.Duration, numBuckets int) *Wheel {
+	// Heuristic to avoid lock contention.
+	numLocks := uint64(runtime.NumCPU() * 2)
+
+	w := &Wheel{
+		period:     period,
+		ticker:     time.NewTicker(period),
+		locks:      make([]*mutex, numLocks),
+		lockMask:   numLocks - 1,
+		buckets:    make([]*Timer, numBuckets),
+		bucketMask: uint64(numBuckets - 1),
+		fired:      make(chan *Timer),
+		stop:       make(chan struct{}),
+		stopped:    make(chan struct{}),
+	}
+	for i := range w.locks {
+		w.locks[i] = &mutex{}
+	}
+	go w.tick()
+	go w.work()
+	return w
+}
+
+// Stop FIXME
+func (w *Wheel) Stop() {
+	w.ticker.Stop()
+	close(w.stop)
+	<-w.stopped
+}
+
+// AfterFunc FIXME
+func (w *Wheel) AfterFunc(d time.Duration, f func()) *Timer {
+	deadline := w.asTick(time.Now().Add(d))
+	t := &Timer{
+		f:        f,
+		deadline: deadline,
+	}
+
+	b := deadline & w.bucketMask
+	mu := w.lockBucket(deadline)
+	w.buckets[b] = w.buckets[b].append(t)
+	mu.Unlock()
+
+	return t
+}
+
+func (w *Wheel) tick() {
+	watermark := w.asTick(time.Now())
+	for {
+		select {
+		case now := <-w.ticker.C:
+			nowTick := w.asTick(now)
+			todo := w.gatherTimers(watermark, nowTick)
+			if todo != nil {
+				w.fired <- todo
+			}
+			watermark = nowTick
+		case <-w.stop:
+			close(w.fired)
+			return
+		}
+	}
+}
+
+func (w *Wheel) work() {
+	for batch := range w.fired {
+		for batch != nil {
+			batch.f()
+			batch = batch.prev
+		}
+	}
+	close(w.stopped)
+}
+
+func (w *Wheel) gatherTimers(startTick, endTick uint64) *Timer {
```

Comment:
If we use a bucket type, and know everything in the bucket has expired, this gets a little simpler too:
 - the wheel just has `[]*bucket`
 - when a tick fires, we swap out the `*bucket` pointer (probably pool it) to a new bucket in the wheel, and then start processing the old bucket
 - processing the old bucket is just iterate over all timers, if the `CAS(stateDefault, stateExpired)` succeeds, then executing the function.
 - once we've executed everything, we reset the bucket `.timers = .timersBacking[:0]` and put it in the bucket pool